### PR TITLE
SCOPE_EXIT helper

### PR DESCRIPTION
**Pull request by Hardcode84**
SCOPE_EXIT helper and usage example in functions.cpp.
My SCOPE_EXIT helper that was discussed in gitter.

### CODE REVIEW

File: gen/functions.cpp [hunk 1]
```diff
@@ -25,6 +25,7 @@
 #include "gen/dvalue.h"
 #include "gen/funcgenstate.h"
 #include "gen/function-inlining.h"
+#include "gen/helpers.h"
```
Comment (ZombineDev) [inline gen/functions.cpp]: `scope_guard.h` would be a more meaningful name.

File: gen/helpers.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+  operator bool() const {
+    return flag;
+  }
+};
+
+template<typename Func>
+struct ScopeExit {
+  Func func;
+  Ownership active = false;
```
Comment (ZombineDev) [inline gen/helpers.h]: Why is this necessary? It seems to me that if you remove it, the class will work the same way.

File: gen/helpers.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  return ScopeExit<typename std::decay<Func>::type>(std::forward<Func>(func));
+}
+
+}
+
+#define LDC_STRINGIZE2(a,b) a##b
+#define LDC_STRINGIZE(a,b) LDC_STRINGIZE2(a,b)
+#define LDC_UNNAME_VAR(basename) LDC_STRINGIZE(basename, __LINE__)
+
+#define SCOPE_EXIT auto LDC_UNNAME_VAR(scope_exit) = details::ScopeExitTag{} << [&]()
```
Comment (ZombineDev) [inline gen/helpers.h]: Personally, I think this going a bit to far, if it's just for the syntax sugar. And I say this, after having watched Alexandrescu's talk.

File: gen/helpers.h (hunk 2 above)
Comment (Hardcode84) [inline gen/helpers.h]: We are returning ScopeExit struct from operator<< function and there can be a temporary copy which shouldn't call lambda.

File: gen/helpers.h (hunk 3 above)
Comment (JohanEngelen) [inline gen/helpers.h]: I also think this is going too far, but am OK with keeping it.

File: gen/functions.cpp (hunk 1 above)
Comment (JohanEngelen) [inline gen/functions.cpp]: @Hardcode84 Can you rename the file to `scope_exit.h`?  (or `scope_guard` if you change the class name too)    I'll merge after.

Comment (Hardcode84) [inline gen/functions.cpp]: Done

**Comment by JohanEngelen** [comment]
Thanks @Hardcode84
